crates/subspace-service/src/sync_from_dsn/fast_sync.rs

```diff
@@ -0,0 +1,587 @@
+        sync_service: Arc<SyncingService<Block>>,
+    ) -> Self {
+        Self {
+            segment_headers_store,
+            node,
+            piece_getter,
+            client,
+            import_queue_service,
+            network_service,
+            fast_sync_state,
+            sync_service,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Sync algorithm:
+    /// - 1. download two last segments,
+    /// - 2. add the last block the second last segment (as raw - without checks and execution),
+    /// - 3. download the state for the first block of the last segment,
+    ///     - add the first block of the last segment as raw,
+    ///     - download state for it,
+    /// - 4. add blocks with execution from the last segment,
+    ///     - notify Substrate sync by importing block with using BlockOrigin::NetworkBroadcast,
+    ///     - clear the block gap to prevent Substrate sync to download blocks from the start.
+    ///     - restart the Substrate sync with SyncMode::Full to enable it downloading full blocks.
+    ///  Note: retry for fast sync will degrade reputation if we have already announced blocks.
+    pub(crate) async fn sync(&self) -> Result<FastSyncResult<Block>, Error> {
+        debug!("Starting fast sync...");
+
+        // 1. Download the last two segments.
+        let mut reconstructor = Reconstructor::new().map_err(|error| error.to_string())?;
+        let segment_header_downloader = SegmentHeaderDownloader::new(self.node);
+
+        if let Err(error) = self
+            .download_segment_headers(&segment_header_downloader)
+            .await
+        {
+            let error = format!("Failed to download segment headers: {}", error);
+            return Err(Error::Other(error));
+        };
+
+        let Some(last_segment_index) = self.segment_headers_store.max_segment_index() else {
+            return Err(Error::Other("Can't get last segment index.".into()));
+        };
+
+        // Skip the fast sync if we lack the minimum required segment number
+        if last_segment_index <= SegmentIndex::ONE {
+            return Ok(FastSyncResult::skipped());
+        }
+
+        let last_segment_header = self
+            .segment_headers_store
+            .get_segment_header(last_segment_index)
+            .expect("Last segment index exists; qed");
+
+        let second_last_segment_index = last_segment_header.segment_index() - SegmentIndex::ONE;
+
+        let second_last_segment_blocks = download_and_reconstruct_blocks(
+            second_last_segment_index,
+            self.piece_getter,
+            &mut reconstructor,
+        )
+        .await?;
+
+        let blocks_in_second_last_segment = second_last_segment_blocks.len();
+
+        if blocks_in_second_last_segment < 2 {
+            return Err(Error::Other(
+                "Unexpected block array length for the second last segment.".into(),
+            ));
+        }
```

Comment:
I think I understand now what you meant to do here, it is just code that doesn't actually talk about it and doesn't check the actual invariant you care about.

The only thing (an nothing else) that we actually care about in fast sync is to import the first block of the last archived segment, this is why spec talks about that and not other things. There is no requirement for 2 blocks to be present, especially not in second last segment.

The edge-case though might be that we could hypothetically have (not practically, but it is still worth supporting IMHO) a block larger that a segment. In that case we may be unable to reconstruct the first block of last segment even if we download two last segments.

So what the code should actually is very close to what spec says:
* once segment headers are known check the header of second last segment and see if last block there was archived or partially
* if fully we only need to download the last segment and nothing else
* if we have a partial block though then we need to check if third last segment ended with the same partial block number because if it did then we need to keep looking forward to the beginning of that block such that we can get the full block of the last segment header regardless of where it starts

With above description it should be clear that this check doesn't make sense. For example it will fail in the perfectly fine situation where second last segment contains end of one block and beginning of another (first block of last segment). In this case we have 0 blocks reconstructed from second last segment alone, yet we are able to successfully reconstruct first block of last segment that the specification actually requires.